components/suggest/src/suggestion.rs

```diff
@@ -31,3 +46,184 @@ impl Suggestion {
         matches!(self, Self::Amp { .. })
     }
 }
+
+/// Replaces all template parameters in a "raw" sponsored suggestion URL,
+/// producing a "cooked" URL with real values.
+pub(crate) fn cook_raw_suggestion_url(raw_url: &str) -> String {
+    let now = Local::now();
+    let mut cooked_url = String::new();
+    let mut last_index = 0;
+    for (index, _) in raw_url.match_indices(TIMESTAMP_TEMPLATE) {
+        cooked_url.push_str(&raw_url[last_index..index]);
+        write!(
+            &mut cooked_url,
+            "{:04}{:02}{:02}{:02}",
+            now.year(),
+            now.month(),
+            now.day(),
+            now.hour()
+        )
+        .expect("Failed to replace timestamp template parameter");
+        last_index = index + TIMESTAMP_TEMPLATE.len();
+    }
+    cooked_url.push_str(&raw_url[last_index..]);
+    cooked_url
+}
+
+/// Determines whether a "raw" sponsored suggestion URL is equivalent to a
+/// "cooked" URL. The two URLs are equivalent if they are identical except for
+/// their replaced template parameters, which can be different.
+pub fn raw_suggestion_url_matches(raw_url: &str, cooked_url: &str) -> bool {
```

Comment:
Likewise, if we leverage the invariant of at-most-1-template-for-each-url, this can be simplified as well.